library/x509_crt.c

```diff
@@ -2681,6 +2681,62 @@ static int x509_crt_verify_restartable_ca_cb( mbedtls_x509_crt *crt,
     return( 0 );
 }
 
+/*
+ * Sort the server cert chain o handle potentially extraneous certificates
+ * and arbitrary orderings.
+ * From https://tools.ietf.org/id/draft-ietf-tls-tls13-23.html#certificate
+ * Prior to TLS 1.3, "certificate_list" ordering required each certificate
+ * to certify the one immediately preceding it; however, some implementations
+ * allowed some flexibility. Servers sometimes send both a current and
+ * deprecated intermediate for transitional purposes, and others are simply
+ * configured incorrectly, but these cases can nonetheless be validated
+ * properly. For maximum compatibility, all implementations SHOULD be prepared
+ * to handle potentially extraneous certificates and arbitrary orderings from
+ * any TLS version, with the exception of the end-entity certificate which
+ * MUST be first.
+ */
+static void mbedtls_x509_crt_sort( mbedtls_x509_crt *chain )
+{
+    mbedtls_x509_crt *child = chain;
+
+    while ( child != NULL)
```

Comment:
Please explain why this loop always terminates. This is not obvious since `child` can be updated in two different ways.

The loop terminates because every run through the loop always removes the current `child` from consideration.

Also, `chain` remains a pointer to the head of the list, because the head is never swapped with some other element.